### ISSUE TITLE
Any plans to add support of JetStream instead of NATS Streaming?

### ISSUE THREAD

**Issue by egorsmkv** [ThreeDotsLabs/watermill#248]
Hello to all contributors!

I am using watermill in my VoIP tech stack to make thousands calls per day and it works perfectly. 
Thanks to all who made this awesome project!

The NATS documentation says the following:
![image](https://user-images.githubusercontent.com/7875085/140582463-5f75886a-3613-4399-9530-5df422a493c0.png)

So my question is simple - Any plans to add support of JetStream instead of NATS Streaming?

**Comment by AlexCuse** [comment]
Been working off and on with @Sacro on this since the summer.  It's pretty slow going but I have some time right now and hoping to get my current WIP in shape to push sometime in the next couple weeks.

**Comment by AlexCuse** [comment]
In general my impression is the streaming package should be moved to its own repo and deprecated since it uses the stan client.

Nats and jet stream should live in watermill-nats since they both will now use the nats client.

**Comment by AlexCuse** [comment]
@egorsmkv have not tested this with a real workload and have a couple things I want to figure out how to clean up but its passing for the following test configuration:

````
tests.Features{
			ConsumerGroups:                      true,
			ExactlyOnceDelivery:                 false,
			GuaranteedOrder:                     true,
			GuaranteedOrderWithSingleSubscriber: true,
			Persistent:                          true,
			RestartServiceCommand:               []string{"docker", "restart", "watermill-jetstream_nats_1"},
			RequireSingleInstance:               false,
			NewSubscriberReceivesOldMessages:    false,
		}
````
		
Give it a try if you get a chance https://github.com/AlexCuse/watermill-jetstream

**Comment by AlexCuse** [comment]
PS I also didn't look at configuration much at all yet but there is a "bring your own connection" factory method that should work OK for whatever you need.

**Comment by egorsmkv** [comment]
Many thanks to you! 

I'll test it and give you my feedback.

**Comment by AlexCuse** [comment]
@egorsmkv I have not been able to get the stress tests to pass but this may be limitations in my ability to set up NATS in docker as much as anything else.  I am primarily using on edge devices that each end up receiving a relatively low message volume and it seems to work great for that.  

Planning to make another pass over the configuration this week and try to tighten that up before I push a 0.0.1 tag just to have something stable to build other packages against.  Let me know if theres anything you know you need on that front.

**Comment by egorsmkv** [comment]
I haven't able to test it yet. Hope to find free time to test it in near time.

**Comment by AlexCuse** [comment]
No rush - I am looking to integrate the KV store with the SDK im working with right now.  Will try and hold off till you have a look before cutting a tag.

**Comment by AlexCuse** [comment]
Hey @roblaszczak - I also realized an ordering issue with a select that could read to 'send on closed channel' panics at the same time so hard to say it was the only thing wrong but I increased the file limit yesterday and seeing stress tests pass consistently now.  Reconnect tests are the next hurdle for testing - maybe the last one - but they should be running as part of the stress test now and this allows other tests to pass with -race.  

One question I find myself with is if it is actually possible for TestMessageCtx (https://github.com/ThreeDotsLabs/watermill/blob/v1.2.0-rc.9/pubsub/tests/test_pubsub.go#L884) to pass with exactly once delivery, at least as nats has envisioned it.  I am using the msg.UUID with `nats.MsgId` and it seems like that can't pass because the broker will never deliver the second message if it is working correctly.
